README.md
<<<<<<< HEAD
:warning::warning:

THIS IS A COPY OF PAPERTRAIL 14.0

This is being used for migrating to a new version of PaperTrail in `refi` only

DO NOT USE

:warning::warning:

# MotorefiPaperTrail

[![Build Status][4]][5]
[![Gem Version][53]][54]
[![SemVer][55]][56]

Track changes to your models, for auditing or versioning. See how a model looked
at any stage in its lifecycle, revert it to any version, or restore it after it
has been destroyed.

## Documentation

This is the _user guide_. See also, the
[API reference](https://www.rubydoc.info/gems/motorefi_paper_trail).

Choose version:
[Unreleased](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/master/README.md),
[14.0](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/v14.0.0/README.md),
[13.0](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/v13.0.0/README.md),
[12.3](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/v12.3.0/README.md),
[11.1](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/v11.1.0/README.md),
[10.3](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/v10.3.1/README.md),
[9.2](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/v9.2.0/README.md),
[8.1](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/v8.1.2/README.md),
[7.1](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/v7.1.3/README.md),
[6.0](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/v6.0.2/README.md),
[5.2](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/v5.2.3/README.md),
[4.2](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/v4.2.0/README.md),
[3.0](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/v3.0.9/README.md),
[2.7](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/v2.7.2/README.md),
[1.6](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/blob/v1.6.5/README.md)

## Table of Contents

<!-- toc -->

- [1. Introduction](#1-introduction)
  - [1.a. Compatibility](#1a-compatibility)
  - [1.b. Installation](#1b-installation)
  - [1.c. Basic Usage](#1c-basic-usage)
  - [1.d. API Summary](#1d-api-summary)
  - [1.e. Configuration](#1e-configuration)
- [2. Limiting What is Versioned, and When](#2-limiting-what-is-versioned-and-when)
  - [2.a. Choosing Lifecycle Events To Monitor](#2a-choosing-lifecycle-events-to-monitor)
  - [2.b. Choosing When To Save New Versions](#2b-choosing-when-to-save-new-motorefi_versions)
  - [2.c. Choosing Attributes To Monitor](#2c-choosing-attributes-to-monitor)
  - [2.d. Turning MotorefiPaperTrail Off](#2d-turning-papertrail-off)
  - [2.e. Limiting the Number of Versions Created](#2e-limiting-the-number-of-motorefi_versions-created)
- [3. Working With Versions](#3-working-with-motorefi_versions)
  - [3.a. Reverting And Undeleting A Model](#3a-reverting-and-undeleting-a-model)
  - [3.b. Navigating Versions](#3b-navigating-motorefi_versions)
  - [3.c. Diffing Versions](#3c-diffing-motorefi_versions)
  - [3.d. Deleting Old Versions](#3d-deleting-old-motorefi_versions)
  - [3.e. Queries](#3e-queries)
  - [3.f. Defunct `item_id`s](#3f-defunct-item_ids)
- [4. Saving More Information About Versions](#4-saving-more-information-about-motorefi_versions)
  - [4.a. Finding Out Who Was Responsible For A Change](#4a-finding-out-who-was-responsible-for-a-change)
  - [4.b. Associations](#4b-associations)
  - [4.c. Storing Metadata](#4c-storing-metadata)
- [5. ActiveRecord](#5-activerecord)
  - [5.a. Single Table Inheritance (STI)](#5a-single-table-inheritance-sti)
  - [5.b. Configuring the `motorefi_versions` Association](#5b-configuring-the-motorefi_versions-association)
  - [5.c. Generators](#5c-generators)
  - [5.d. Protected Attributes](#5d-protected-attributes)
- [6. Extensibility](#6-extensibility)
  - [6.a. Custom Version Classes](#6a-custom-version-classes)
  - [6.b. Custom Serializer](#6b-custom-serializer)
  - [6.c. Custom Object Changes](#6c-custom-object-changes)
  - [6.d. Excluding the Object Column](#6d-excluding-the-object-column)
- [7. Testing](#7-testing)
  - [7.a. Minitest](#7a-minitest)
  - [7.b. RSpec](#7b-rspec)
  - [7.c. Cucumber](#7c-cucumber)
  - [7.d. Spork](#7d-spork)
  - [7.e. Zeus or Spring](#7e-zeus-or-spring)
- [8. MotorefiPaperTrail Plugins](#8-papertrail-plugins)
- [9. Integration with Other Libraries](#9-integration-with-other-libraries)
- [10. Related Libraries and Ports](#10-related-libraries-and-ports)
- [Articles](#articles)
- [Problems](#problems)
- [Contributors](#contributors)
- [Contributing](#contributing)
- [Inspirations](#inspirations)
- [Intellectual Property](#intellectual-property)

<!-- tocstop -->

## 1. Introduction

### 1.a. Compatibility

| motorefi_paper_trail | branch     | ruby     | activerecord  |
| -------------------- | ---------- | -------- | ------------- |
| unreleased           | master     | >= 2.7.0 | >= 6.0, < 7.1 |
| 14                   | 14-stable  | >= 2.7.0 | >= 6.0, < 7.1 |
| 13                   | 13-stable  | >= 2.6.0 | >= 5.2, < 7.1 |
| 12                   | 12-stable  | >= 2.6.0 | >= 5.2, < 7.1 |
| 11                   | 11-stable  | >= 2.4.0 | >= 5.2, < 6.1 |
| 10                   | 10-stable  | >= 2.3.0 | >= 4.2, < 6.1 |
| 9                    | 9-stable   | >= 2.3.0 | >= 4.2, < 5.3 |
| 8                    | 8-stable   | >= 2.2.0 | >= 4.2, < 5.2 |
| 7                    | 7-stable   | >= 2.1.0 | >= 4.0, < 5.2 |
| 6                    | 6-stable   | >= 1.9.3 | >= 4.0, < 5.2 |
| 5                    | 5-stable   | >= 1.9.3 | >= 3.0, < 5.1 |
| 4                    | 4-stable   | >= 1.8.7 | >= 3.0, < 5.1 |
| 3                    | 3.0-stable | >= 1.8.7 | >= 3.0, < 5   |
| 2                    | 2.7-stable | >= 1.8.7 | >= 3.0, < 4   |
| 1                    | rails2     | >= 1.8.7 | >= 2.3, < 3   |

Experts: to install incompatible motorefi_versions of activerecord, see
`motorefi_paper_trail/compatibility.rb`.

### 1.b. Installation

1.  Add MotorefiPaperTrail to your `Gemfile` and run [`bundle`][57].

    `gem 'motorefi_paper_trail'`

1.  Add a `motorefi_versions` table to your database:

    ```
    bundle exec rails generate motorefi_paper_trail:install [--with-changes]
    ```

    If tables in your project use `uuid` instead of `integers` for `id`, then use:

    ```
    bundle exec rails generate motorefi_paper_trail:install [--uuid]
    ```

    See [section 5.c. Generators](#5c-generators) for details.

    ```
    bundle exec rake db:migrate
    ```

1.  Add `has_motorefi_paper_trail` to the models you want to track.

    ```ruby
    class Widget < ActiveRecord::Base
      has_motorefi_paper_trail
    end
    ```

1.  If your controllers have a `current_user` method, you can easily [track who
    is responsible for changes](#4a-finding-out-who-was-responsible-for-a-change)
    by adding a controller callback.

        ```ruby
        class ApplicationController
          before_action :set_motorefi_paper_trail_whodunnit
        end
        ```

### 1.c. Basic Usage

Your models now have a `motorefi_versions` method which returns the "paper trail" of
changes to your model.

```ruby
widget = Widget.find 42
widget.motorefi_versions
# [<MotorefiPaperTrail::Version>, <MotorefiPaperTrail::Version>, ...]
```

Once you have a version, you can find out what happened:

```ruby
v = widget.motorefi_versions.last
v.event # 'update', 'create', 'destroy'. See also: "The motorefi_versions.event Column"
v.created_at
v.whodunnit # ID of `current_user`. Requires `set_motorefi_paper_trail_whodunnit` callback.
widget = v.reify # The widget as it was before the update (nil for a create event)
```

MotorefiPaperTrail stores the pre-change version of the model, unlike some other
auditing/versioning plugins, so you can retrieve the original version. This is
useful when you start keeping a paper trail for models that already have records
in the database.

```ruby
widget = Widget.find 153
widget.name                                 # 'Doobly'

# Add has_motorefi_paper_trail to Widget model.

widget.motorefi_versions                             # []
widget.update name: 'Wotsit'
widget.motorefi_versions.last.reify.name             # 'Doobly'
widget.motorefi_versions.last.event                  # 'update'
```

This also means that MotorefiPaperTrail does not waste space storing a version of the
object as it currently stands. The `motorefi_versions` method gives you previous
motorefi_versions; to get the current one just call a finder on your `Widget` model as
usual.

Here's a helpful table showing what MotorefiPaperTrail stores:

| _Event_        | _create_ | _update_ | _destroy_ |
| -------------- | -------- | -------- | --------- |
| _Model Before_ | nil      | widget   | widget    |
| _Model After_  | widget   | widget   | nil       |

MotorefiPaperTrail stores the values in the Model Before row. Most other
auditing/versioning plugins store the After row.

### 1.d. API Summary

An introductory sample of common features.

When you declare `has_motorefi_paper_trail` in your model, you get these methods:

```ruby
class Widget < ActiveRecord::Base
  has_motorefi_paper_trail
end

# Returns this widget's motorefi_versions.  You can customise the name of the
# association, but overriding this method is not supported.
widget.motorefi_versions

# Return the version this widget was reified from, or nil if it is live.
# You can customise the name of the method.
widget.version

# Returns true if this widget is the current, live one; or false if it is from
# a previous version.
widget.motorefi_paper_trail.live?

# Returns who put the widget into its current state.
widget.motorefi_paper_trail.originator

# Returns the widget (not a version) as it looked at the given timestamp.
widget.motorefi_paper_trail.version_at(timestamp)

# Returns the widget (not a version) as it was most recently.
widget.motorefi_paper_trail.previous_version

# Returns the widget (not a version) as it became next.
widget.motorefi_paper_trail.next_version
```

And a `MotorefiPaperTrail::Version` instance (which is just an ordinary ActiveRecord
instance, with all the usual methods) has methods such as:

```ruby
# Returns the item restored from this version.
version.reify(options = {})

# Return a new item from this version
version.reify(dup: true)

# Returns who put the item into the state stored in this version.
version.motorefi_paper_trail_originator

# Returns who changed the item from the state it had in this version.
version.terminator
version.whodunnit
version.version_author

# Returns the next version.
version.next

# Returns the previous version.
version.previous

# Returns the index of this version in all the motorefi_versions.
version.index

# Returns the event that caused this version (create|update|destroy).
version.event
```

This is just a sample of common features. Keep reading for more.

### 1.e. Configuration

Many aspects of MotorefiPaperTrail are configurable for individual models; typically
this is achieved by passing options to the `has_motorefi_paper_trail` method within
a given model.

Some aspects of MotorefiPaperTrail are configured globally for all models. These
settings are assigned directly on the `MotorefiPaperTrail.config` object.
A common place to put these settings is in a Rails initializer file
such as `config/initializers/motorefi_paper_trail.rb` or in an environment-specific
configuration file such as `config/environments/test.rb`.

#### 1.e.1 Global

Global configuration options affect all threads.

- association_reify_error_behaviour
- enabled
- has_motorefi_paper_trail_defaults
- object_changes_adapter
- serializer
- version_limit

Syntax example: (options described in detail later)

```ruby
# config/initializers/motorefi_paper_trail.rb
MotorefiPaperTrail.config.enabled = true
MotorefiPaperTrail.config.has_motorefi_paper_trail_defaults = {
  on: %i[create update destroy]
}
MotorefiPaperTrail.config.version_limit = 3
```

These options are intended to be set only once, during app initialization (eg.
in `config/initializers`). It is unsafe to change them while the app is running.
In contrast, `MotorefiPaperTrail.request` has various options that only apply to a
single HTTP request and thus are safe to use while the app is running.

## 2. Limiting What is Versioned, and When

### 2.a. Choosing Lifecycle Events To Monitor

You can choose which events to track with the `on` option. For example, if
you only want to track `update` events:

```ruby
class Article < ActiveRecord::Base
  has_motorefi_paper_trail on: [:update]
end
```

`has_motorefi_paper_trail` installs [callbacks][52] for the specified lifecycle events.

There are four potential callbacks, and the default is to install all four, ie.
`on: [:create, :destroy, :touch, :update]`.

#### The `motorefi_versions.event` Column

Your `motorefi_versions` table has an `event` column with three possible values:

| _event_ | _callback_    |
| ------- | ------------- |
| create  | create        |
| destroy | destroy       |
| update  | touch, update |

You may also have the `MotorefiPaperTrail::Version` model save a custom string in its
`event` field instead of the typical `create`, `update`, `destroy`. MotorefiPaperTrail
adds an `attr_accessor` to your model named `motorefi_paper_trail_event`, and will insert
it, if present, in the `event` column.

```ruby
a = Article.create
a.motorefi_versions.size                           # 1
a.motorefi_versions.last.event                     # 'create'
a.motorefi_paper_trail_event = 'update title'
a.update title: 'My Title'
a.motorefi_versions.size                           # 2
a.motorefi_versions.last.event                     # 'update title'
a.motorefi_paper_trail_event = nil
a.update title: 'Alternate'
a.motorefi_versions.size                           # 3
a.motorefi_versions.last.event                     # 'update'
```

#### Controlling the Order of AR Callbacks

If there are other callbacks in your model, their order relative to those
installed by `has_motorefi_paper_trail` may matter. If you need to control
their order, use the `motorefi_paper_trail_on_*` methods.

```ruby
class Article < ActiveRecord::Base
  # Include MotorefiPaperTrail, but do not install any callbacks. Passing the
  # empty array to `:on` omits callbacks.
  has_motorefi_paper_trail on: []

  # Add callbacks in the order you need.
  motorefi_paper_trail.on_destroy    # add destroy callback
  motorefi_paper_trail.on_update     # etc.
  motorefi_paper_trail.on_create
  motorefi_paper_trail.on_touch
end
```

The `motorefi_paper_trail.on_destroy` method can be further configured to happen
`:before` or `:after` the destroy event. Until MotorefiPaperTrail 4, the default was
`:after`. Starting with MotorefiPaperTrail 5, the default is `:before`, to support
ActiveRecord 5. (see https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/683)

### 2.b. Choosing When To Save New Versions

You can choose the conditions when to add new motorefi_versions with the `if` and
`unless` options. For example, to save motorefi_versions only for US non-draft
translations:

```ruby
class Translation < ActiveRecord::Base
  has_motorefi_paper_trail if:     Proc.new { |t| t.language_code == 'US' },
                  unless: Proc.new { |t| t.type == 'DRAFT'       }
end
```

#### Choosing Based on Changed Attributes

Starting with MotorefiPaperTrail 4.0, motorefi_versions are saved during an after-callback. If
you decide whether to save a new version based on changed attributes,
use attribute_name_was instead of attribute_name.

#### Saving a New Version Manually

You may want to save a new version regardless of options like `:on`, `:if`, or
`:unless`. Or, in rare situations, you may want to save a new version even if
the record has not changed.

```ruby
my_model.motorefi_paper_trail.save_with_version
```

### 2.c. Choosing Attributes To Monitor

#### Ignore

If you don't want a version created when only a certain attribute changes, you can `ignore` that attribute:

```ruby
class Article < ActiveRecord::Base
  has_motorefi_paper_trail ignore: [:title, :rating]
end
```

Changes to just the `title` or `rating` will not create a version record.
Changes to other attributes will create a version record.

```ruby
a = Article.create
a.motorefi_versions.length                         # 1
a.update title: 'My Title', rating: 3
a.motorefi_versions.length                         # 1
a.update title: 'Greeting', content: 'Hello'
a.motorefi_versions.length                         # 2
a.motorefi_paper_trail.previous_version.title      # 'My Title'
```

Note: ignored fields will be stored in the version records. If you want to keep a field out of the motorefi_versions table, use [`:skip`](#skip) instead of `:ignore`; skipped fields are also implicitly ignored.

The `:ignore` option can also accept `Hash` arguments that we are considering deprecating.

```ruby
class Article < ActiveRecord::Base
  has_motorefi_paper_trail ignore: [:title, { color: proc { |obj| obj.color == "Yellow" } }]
end
```

#### Only

Or, you can specify a list of the `only` attributes you care about:

```ruby
class Article < ActiveRecord::Base
  has_motorefi_paper_trail only: [:title]
end
```

Only changes to the `title` will create a version record.

```ruby
a = Article.create
a.motorefi_versions.length                         # 1
a.update title: 'My Title'
a.motorefi_versions.length                         # 2
a.update content: 'Hello'
a.motorefi_versions.length                         # 2
a.motorefi_paper_trail.previous_version.content    # nil
```

The `:only` option can also accept `Hash` arguments that we are considering deprecating.

```ruby
class Article < ActiveRecord::Base
  has_motorefi_paper_trail only: [{ title: Proc.new { |obj| !obj.title.blank? } }]
end
```

If the `title` is not blank, then only changes to the `title`
will create a version record.

```ruby
a = Article.create
a.motorefi_versions.length                         # 1
a.update content: 'Hello'
a.motorefi_versions.length                         # 2
a.update title: 'Title One'
a.motorefi_versions.length                         # 3
a.update content: 'Hai'
a.motorefi_versions.length                         # 3
a.motorefi_paper_trail.previous_version.content    # "Hello"
a.update title: 'Title Two'
a.motorefi_versions.length                         # 4
a.motorefi_paper_trail.previous_version.content    # "Hai"
```

Configuring both `:ignore` and `:only` is not recommended, but it should work as
expected. Passing both `:ignore` and `:only` options will result in the
article being saved if a changed attribute is included in `:only` but not in
`:ignore`.

#### Skip

If you never want a field's values in the motorefi_versions table, you can `:skip` the attribute. As with `:ignore`,
updates to these attributes will not create a version record. In addition, if a
version record is created for some other reason, these attributes will not be
persisted.

```ruby
class Author < ActiveRecord::Base
  has_motorefi_paper_trail skip: [:social_security_number]
end
```

Author's social security numbers will never appear in the motorefi_versions log, and if an author updates only their social security number, it won't create a version record.

#### Comparing `:ignore`, `:only`, and `:skip`

- `:only` is basically the same as `:ignore`, but its inverse.
- `:ignore` controls whether motorefi_paper_trail will create a version record or not.
- `:skip` controls whether motorefi_paper_trail will save that field with the version record.
- Skipped fields are also implicitly ignored. motorefi_paper_trail does this internally.
- Ignored fields are not implicitly skipped.

So:

- Ignore a field if you don't want a version record created when it's the only field to change.
- Skip a field if you don't want it to be saved with any version records.

### 2.d. Turning MotorefiPaperTrail Off

MotorefiPaperTrail is on by default, but sometimes you don't want to record motorefi_versions.

#### Per Process

Turn MotorefiPaperTrail off for **all threads** in a `ruby` process.

```ruby
MotorefiPaperTrail.enabled = false
```

**Do not use this in production** unless you have a good understanding of
threads vs. processes.

A legitimate use case is to speed up tests. See [Testing](#7-testing) below.

#### Per HTTP Request

```ruby
MotorefiPaperTrail.request(enabled: false) do
  # no motorefi_versions created
end
```

or,

```ruby
MotorefiPaperTrail.request.enabled = false

CHANGELOG.md
# Changelog

This project follows [semver 2.0.0](http://semver.org/spec/v2.0.0.html) and the
recommendations of [keepachangelog.com](http://keepachangelog.com/).

## Unreleased

### Breaking Changes

- None

### Added

- None

### Fixed

- None

## 14.0.0 (2022-11-26)

### Breaking Changes

- [#1399](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1399) - Same
  change re: `YAML.safe_load` as in 13.0.0, but this time for Rails 6.0 and 6.1.
- [#1406](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1406) -
  Certain [Metadata][1] keys are now forbidden, like `id`, and `item_type`.
  These keys are reserved by PT.

### Dependencies

- Drop support for Rails 5.2, which reached EoL on 2022-06-01
- Drop support for Ruby 2.6, which reached EoL on 2022-03-31
- Drop support for request_store < 1.4

### Added

- None

### Fixed

- [#1395](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/issues/1395) -
  Fix incorrect `Version#created_at` value when using
  `MotorefiPaperTrail::RecordTrail#update_columns`
- [#1404](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1404) -
  Delay referencing ActiveRecord until after Railtie is loaded
- Where possible, methods which are not part of MotorefiPaperTrail's public API have
  had their access changed to private. All of these methods had been clearly
  marked as `@api private` in the documentation, for years. This is not expected
  to be a breaking change.

## 13.0.0 (2022-08-15)

### Breaking Changes

- For Rails >= 7.0, the default serializer will now use `YAML.safe_load` unless
  `ActiveRecord.use_yaml_unsafe_load`. This change only affects users whose
  `motorefi_versions` table has `object` or `object_changes` columns of type `text`, and
  who use the YAML serializer. People who use the JSON serializer, or those with
  `json(b)` columns, are unaffected. Please see
  [doc/pt_13_yaml_safe_load.md](doc/pt_13_yaml_safe_load.md) for details.

### Added

- None

### Fixed

- None

## 12.3.0 (2022-03-13)

### Breaking Changes

- None

### Added

- [#1371](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1371) - Added
  `in_after_callback` argument to `MotorefiPaperTrail::RecordTrail#save_with_version`,
  to allow the caller to indicate if this method is being called during an
  `after` callback. Defaults to `false`.
- [#1374](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1374) - Added
  option `--uuid` when generating new migration. This can be used to set the
  type of item_id column to uuid for use with motorefi_paper_trail on a database that
  uses uuid as primary key.

### Fixed

- [#1373](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/issues/1373) - Add
  CLI option to use uuid type for item_id when generating migration.
- [#1376](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1376) - Create a
  version record when associated object is touched. Restores the behavior of
  MotorefiPaperTrail < v12.1.0.

## 12.2.0 (2022-01-21)

### Breaking Changes

- None

### Added

- [#1365](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1365) -
  Support Rails 7.0
- [#1349](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1349) -
  `if:` and `unless:` work with `touch` events now.

### Fixed

- [#1366](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1366) -
  Fixed a bug where the `create_versions` migration lead to a broken `db/schema.rb` for Ruby 3

### Dependencies

- [#1338](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1338) -
  Support Psych version 4
- ruby >= 2.6 (was >= 2.5). Ruby 2.5 reached EoL on 2021-03-31.

## 12.1.0 (2021-08-30)

### Breaking Changes

- None

### Added

- [#1292](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1292) -
  `where_attribute_changes` queries for motorefi_versions where the object's attribute
  changed to or from any values.
- [#1291](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1291) -
  `where_object_changes_to` queries for motorefi_versions where the object's attributes
  changed to one set of known values from any other set of values.

### Fixed

- [#1285](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1285) -
  For ActiveRecord >= 6.0, the `touch` callback will no longer create a new
  `Version` for skipped or ignored attributes.
- [#1309](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1309) -
  Removes `item_subtype` requirement when specifying model-specific limits.
- [#1333](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1333) -
  Improve reification of STI models that use `find_sti_class`/`sti_class_for`
  to customize single table inheritance.

## 12.0.0 (2021-03-29)

### Breaking Changes

- [#1281](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1281) Rails:
  Instead of an `Engine`, PT now provides a `Railtie`, which is simpler.
  This was not expected to be a breaking change, but has caused trouble for
  some people:
  - Issue with the deprecated `autoloader = :classic` setting
    (https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/issues/1305)
- Rails: The deprecated `config.motorefi_paper_trail` configuration technique
  has been removed. This configuration object was deprecated in 10.2.0. It only
  had one key, `config.motorefi_paper_trail.enabled`. Please review docs section [2.d.
  Turning MotorefiPaperTrail
  Off](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/#2d-turning-papertrail-off)
  for alternatives.

### Added

- `where_object_changes_from` queries for motorefi_versions where the object's attributes
  changed from one set of known values to any other set of values.

### Fixed

- [#1281](https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1281) Rails:

lib/motorefi_paper_trail/model_config.rb
# frozen_string_literal: true

module MotorefiPaperTrail
  # Configures an ActiveRecord model, mostly at application boot time, but also
  # sometimes mid-request, with methods like enable/disable.
  class ModelConfig
    E_CANNOT_RECORD_AFTER_DESTROY = <<-STR.strip_heredoc.freeze
      motorefi_paper_trail.on_destroy(:after) is incompatible with ActiveRecord's
      belongs_to_required_by_default. Use on_destroy(:before)
      or disable belongs_to_required_by_default.
    STR
    E_HPT_ABSTRACT_CLASS = <<~STR.squish.freeze
      An application model (%s) has been configured to use MotorefiPaperTrail (via
      `has_motorefi_paper_trail`), but the version model it has been told to use (%s) is
      an `abstract_class`. This could happen when an advanced feature called
      Custom Version Classes (http://bit.ly/2G4ch0G) is misconfigured. When all
      version classes are custom, MotorefiPaperTrail::Version is configured to be an
      `abstract_class`. This is fine, but all application models must be
      configured to use concrete (not abstract) version models.
    STR
    DPR_PASSING_ASSOC_NAME_DIRECTLY_TO_VERSIONS_OPTION = <<~STR.squish
      Passing motorefi_versions association name as `has_motorefi_paper_trail motorefi_versions: %{versions_name}`
      is deprecated. Use `has_motorefi_paper_trail motorefi_versions: {name: %{versions_name}}` instead.
      The hash you pass to `motorefi_versions:` is now passed directly to `has_many`.
    STR
    DPR_CLASS_NAME_OPTION = <<~STR.squish
      Passing Version class name as `has_motorefi_paper_trail class_name: %{class_name}`
      is deprecated. Use `has_motorefi_paper_trail motorefi_versions: {class_name: %{class_name}}`
      instead. The hash you pass to `motorefi_versions:` is now passed directly to `has_many`.
    STR

    def initialize(model_class)
      @model_class = model_class
    end

    # Adds a callback that records a version after a "create" event.
    #
    # @api public
    def on_create
      @model_class.after_create { |r|
        r.motorefi_paper_trail.record_create if r.motorefi_paper_trail.save_version?
      }
      append_option_uniquely(:on, :create)
    end

    # Adds a callback that records a version before or after a "destroy" event.
    #
    # @api public
    def on_destroy(recording_order = "before")
      assert_valid_recording_order_for_on_destroy(recording_order)
      @model_class.send(
        "#{recording_order}_destroy",
        lambda do |r|
          return unless r.motorefi_paper_trail.save_version?
          r.motorefi_paper_trail.record_destroy(recording_order)
        end
      )
      append_option_uniquely(:on, :destroy)
    end

    # Adds a callback that records a version after an "update" event.
    #
    # @api public
    def on_update
      @model_class.before_save { |r|
        r.motorefi_paper_trail.reset_timestamp_attrs_for_update_if_needed
      }
      @model_class.after_update { |r|
        if r.motorefi_paper_trail.save_version?
          r.motorefi_paper_trail.record_update(
            force: false,
            in_after_callback: true,
            is_touch: false
          )
        end
      }
      @model_class.after_update { |r|
        r.motorefi_paper_trail.clear_version_instance
      }
      append_option_uniquely(:on, :update)
    end

    # Adds a callback that records a version after a "touch" event.
    #
    # Rails < 6.0 has a bug where dirty-tracking does not occur during
    # a `touch`. (https://github.com/rails/rails/issues/33429) See also:
    # https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/issues/1121
    # https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/issues/1161
    # https://github.com/motorefi-paper-trail-gem/motorefi_paper_trail/pull/1285
    #
    # @api public
    def on_touch
      @model_class.after_touch { |r|
        if r.motorefi_paper_trail.save_version?
          r.motorefi_paper_trail.record_update(
            force: false,
            in_after_callback: true,
            is_touch: true
          )
        end
      }
    end

    # Set up `@model_class` for MotorefiPaperTrail. Installs callbacks, associations,
    # "class attributes", instance methods, and more.
    # @api private
    def setup(options = {})
      options[:on] ||= %i[create update destroy touch]
      options[:on] = Array(options[:on]) # Support single symbol
      @model_class.send :include, ::MotorefiPaperTrail::Model::InstanceMethods
      setup_options(options)
      setup_associations(options)
      @model_class.after_rollback { motorefi_paper_trail.clear_rolled_back_versions }
      setup_callbacks_from_options options[:on]
    end

    # @api private
    def version_class
      "MotorefiPaperTrail::Version".constantize
    end

    private

    # @api private
    def append_option_uniquely(option, value)
      collection = @model_class.motorefi_paper_trail_options.fetch(option)
      return if collection.include?(value)
      collection << value
    end

    # Raises an error if the provided class is an `abstract_class`.
    # @api private
    def assert_concrete_activerecord_class(class_name)
      if class_name.constantize.abstract_class?
        raise Error, format(E_HPT_ABSTRACT_CLASS, @model_class, class_name)
      end
    end

    # @api private
    def assert_valid_recording_order_for_on_destroy(recording_order)
      unless %w[after before].include?(recording_order.to_s)
        raise ArgumentError, 'recording order can only be "after" or "before"'
      end

      if recording_order.to_s == "after" && cannot_record_after_destroy?
        raise Error, E_CANNOT_RECORD_AFTER_DESTROY
      end
    end

    def cannot_record_after_destroy?
      ::ActiveRecord::Base.belongs_to_required_by_default
    end

    def check_version_class_name(options)
      # @api private - `version_class_name`
      @model_class.class_attribute :version_class_name
      if options[:class_name]
        ::ActiveSupport::Deprecation.warn(
          format(
            DPR_CLASS_NAME_OPTION,
            class_name: options[:class_name].inspect
          ),
          caller(1)
        )
        options[:motorefi_versions][:class_name] = options[:class_name]
      end
      @model_class.version_class_name = options[:motorefi_versions][:class_name] || "MotorefiPaperTrail::Version"
      assert_concrete_activerecord_class(@model_class.version_class_name)
    end

    def check_versions_association_name(options)
      # @api private - versions_association_name
      @model_class.class_attribute :versions_association_name
      @model_class.versions_association_name = options[:motorefi_versions][:name] || :motorefi_versions
    end

    def define_has_many_versions(options)
      options = ensure_versions_option_is_hash(options)
      check_version_class_name(options)
      check_versions_association_name(options)
      scope = get_versions_scope(options)
      @model_class.has_many(
        @model_class.versions_association_name,
        scope,
        class_name: @model_class.version_class_name,
        as: :item,
        **options[:motorefi_versions].except(:name, :scope)
      )
    end

    def ensure_versions_option_is_hash(options)
      unless options[:motorefi_versions].is_a?(Hash)
        if options[:motorefi_versions]
          ::ActiveSupport::Deprecation.warn(
            format(
              DPR_PASSING_ASSOC_NAME_DIRECTLY_TO_VERSIONS_OPTION,
              versions_name: options[:motorefi_versions].inspect
            ),
            caller(1)
          )
        end
        options[:motorefi_versions] = {
          name: options[:motorefi_versions]
        }
      end
      options
    end

    # Process an `ignore`, `skip`, or `only` option.
    def event_attribute_option(option_name)
      [@model_class.motorefi_paper_trail_options[option_name]].
        flatten.
        compact.
        map { |attr| attr.is_a?(Hash) ? attr.stringify_keys : attr.to_s }
    end

    def get_versions_scope(options)
      options[:motorefi_versions][:scope] || -> { order(model.timestamp_sort_order) }
    end

    def setup_associations(options)
      # @api private - version_association_name
      @model_class.class_attribute :version_association_name
      @model_class.version_association_name = options[:version] || :version

      # The version this instance was reified from.
      # @api public
      @model_class.send :attr_accessor, @model_class.version_association_name

      # @api public - motorefi_paper_trail_event
      @model_class.send :attr_accessor, :motorefi_paper_trail_event

      define_has_many_versions(options)
    end

    def setup_callbacks_from_options(options_on = [])
      options_on.each do |event|
        public_send("on_#{event}")
      end
    end

    def setup_options(options)
      # @api public - motorefi_paper_trail_options - Let's encourage plugins to use eg.
      # `motorefi_paper_trail_options[:motorefi_versions][:class_name]` rather than
      # `version_class_name` because the former is documented and the latter is
      # not.
      @model_class.class_attribute :motorefi_paper_trail_options
      @model_class.motorefi_paper_trail_options = options.dup

      %i[ignore skip only].each do |k|
        @model_class.motorefi_paper_trail_options[k] = event_attribute_option(k)
      end
      @model_class.motorefi_paper_trail_options[:meta] ||= {}
    end
  end
end
